text/0000-derefered_composite_types.md

```diff
@@ -0,0 +1,110 @@
+- Feature Name: `derefered_composite_types`
+- Start Date: 2023-04-06
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+
+# Summary
+[summary]: #summary
+
+The proposal of Derefered Composite Types is a Rust unification of Refered Types and Derefered Types.
+
+Derefered Composite Types could help also to get rid of constructor boilerplate.
+
+The symbol `*` before the type name is a marker of Derefered Composite Type. 
+
+The new operator `=&` (and `=&&`, `=&&&`, ...) as a compound of borrow and assignment (`x = &2` same as `x =& 2`) is also required.
+
+
+# Motivation
+[motivation]: #motivation
+
+Currently, Rust has 
+- (1) derefered primitive types (like `i32`)
+- (2) refered primitive types (like `&i32`) 
+- (3) refered composite types (like `Box<i32>`)
```

Comment:
I have never heard of this mental model before. For me, `i32` is a just a type. `&i32` is a reference to a type (the fact that  `i32` is a primitive isn't relevant here). `Box<i32>` is a smart pointer. In this model, there is no gap left to fix.

With my mental model this RFC makes no sense. What does "derefered composite type" even mean? What _is_ a "composite type" for you?

Reply:
Unfortunately, `i32` is a type of Memory link (2 numbers - variable starting allocation and known length) to a Number, but not a Number itself!
Just like `Box<i32>` is.
But both use different syntax to construct and dereferece as I mentioned in RFC.

Reply:
`i32` is definitely a number itself.

Reply:
Ok, my explanations wasn't clear enough.
@Lokathor , yes, you are right, `i32` is a type of Number. But it was not what I tried to explain.
Consider we have 
```rust
let foo : i32 = 5i32;
```
Is this `foo` a pointer to a number? Yes, it is.
Is this `5i32` a number? Yes, it is.
So, is it possible, that `foo` and `5i32` has **the same** type? No, it is not possible!
`5i32` is a `i32` type. And variable `foo` has hidden type `Stack<i32>`, hidden constructor and hidden deref.
But `Box<i32>` has unhidden type, unhidden constructor and unhidden deref.

So I propose to use star-marked types also use hidden constructor and hidden deref

Reply:
@VitWW I'm afraid you seem to have a very incorrect understanding of the topic.

> And variable foo has hidden type Stack<i32>, hidden constructor and hidden deref.

This is completely false. There is no "hidden constructor" or "hidden deref". There is no "hidden type Stack<i32>".

>  But Box<i32> has unhidden type, unhidden constructor and unhidden deref.

There is no special treatment for "box" in this respect. Box is a struct wrapper around a pointer to an `i32`. The pointer itself may be stored on the stack just as a plain `i32` may be stored on the stack. You are assuming some special treatment of `i32` which does not exist.

The RFC is non-sensical because it's based on these falsehoods.

Reply:
But it looks like Rust has this behavior and the Type system does not forbid to think on this level of abstraction.

Reply:
You can imagine Rust a lot of ways, but `i32` is not a pointer to a number, it is the number directly.

If your way of imagining Rust leads you to think that `i32` is something other than the direct number then that way is incorrect.

Reply:
> But it looks like Rust has this behavior and the Type system does not forbid to think on this level of abstraction.

Aside from the fact that the abstraction is wrong and so will lead to incorrect conclusions if you reason using it, Rust does actually explicitly forbid thinking in terms of *this type* of abstraction. That's because Rust is defined in terms of an abstract machine on which Rust code executes, and that abstract machine doesn't operate in terms of addresses on a stack: it simply talks about places and values, and it's unspecified how that relates to actual memory addresses. Those places could end up on the stack, in registers, or wherever else the Rust compiler feels like putting them.

For a real example of this, see here: https://rust.godbolt.org/z/aEoWjWjzs

You can see that despite accepting a parameter, and defining a local variable *nothing* is stored on the stack during execution of this function. Both the parameter and the return value are stored directly in registers, and no dereferencing of pointers ever occurs.

Reply:
@Diggsey Thank you.
It looks that even this RFC is still capable to adding, it has much harder implementation than I expected.
And while it has no **significant** advantages, it remains as an idea only.
Till someone finds useful application.

Reply:
@Diggsey 
It looks like it is possible do not implement derefered types, but still have their smart constructors!!
I prefer _**smart constructors**_ more than types. )))

(A) I wish to have `FromDeref` Trait from the next:
```rust
let *x : Box<i32> = 5;
```
which desugars into
```rust
let x : Box<i32> = Box::from_deref(5);
```
We expect, that `FromDeref` Trait is dual to `Deref` Trait and follow next rule:
```rust
let x  : SomeType;
let *y : SomeType = *x;
```

(B) I also wish to have an ability for `as`cast to use for types, that follow `From` and `Into` Traits
```rust
let s = "text" as String;
```
which desugars into
```rust
let s = String::from("text");
```
Is it sound reasonable proposal?

Reply:
Generalizing `as` may be reasonable, but `FromDeref` looks like it will add complexity to the compiler and in developers' minds without being more expressive. Your example is barely shorter than the code they would replace (especially since it's only on initialization), and the syntax might even be confused as a pointer to a `Box`.

Reply:
In general, the language has been moving away from `as` casting over time, and probably we wouldn't add new ways to use `as` for just a shorthand.

Reply:
Syntax nit: I can understand if you want something like `String("text")` (even though it's not practical due to its collision with the tuple struct constructor syntax), but `"text" as String` is definitely worse, because it will cause even more precedence problems. Every time I use the `as` operator in a language that supports it, I have to look up the documentation to check whether `1 + 2 as Foo` etc. is `(1 + 2) as Foo` or `1 + (2 as Foo)`.

Reply:
@Lokathor `as` casting has limited use now because it is a cost-less casting only (but not from literals). If it has more abilities, it would have much wider use.

@SOF3 I agree, `as` has almost highest precedence (=13), so we also need "weak precedence" synonym-keyword `as'` with precedence =3 (a bit stronger than assignments operators)
So `1 + 2 as Foo as' Bar` mean `(1 + (2 as Foo)) as' Bar`